### ISSUE TITLE
Grub-btrfs doesn't detect my Timeshift btrfs snapshots on running "grub-mkconfig"

### ISSUE THREAD

**Issue by Xe-no1** [Antynea/grub-btrfs#344]
As the title suggests, Grub-btrfs is not detecting my btrfs snapshots (using Timeshift), I configured the Grub-btrfs daemon to monitor the Timeshift snapshots using "--timeshift-auto" but without any results or detecting my snapshots, even though I'm sure I have snapshots. My version of Timeshift is 24.06.3 and Grub-btrfs is version 4.12. Im using vanilla Arch linux with KDE Plasma Wayland on a VM and my host computer is a MacBook Pro 14 inch 2021, meaning everything is aarch64 and my Arch linux is the ARM flavor
This is the exact output I get from running "grub-mkconfig": 
[grub-mkconfig output.txt](https://github.com/user-attachments/files/16447141/grub-mkconfig.output.txt)
These are my grub configs: 
[grub config.txt](https://github.com/user-attachments/files/16447136/grub.config.txt)
These are my grub-btrfs configs: 
[grub-btrfs config.txt](https://github.com/user-attachments/files/16447147/grub-btrfs.config.txt)
These are my grub-btrfs daemon configs: 
[grub-btrfs daemon config.txt](https://github.com/user-attachments/files/16447151/grub-btrfs.daemon.config.txt)
Feel free to ask me to add any needed info that I missed, I am **_really_** **_really_** new to Linux (literally started 2 weeks ago) and I try to add any relevant info that I know of.

**Comment by DZve** [comment]
Try remaking grub, the automation might not be working. 
`grub-mkconfig -o /boot/grub/grub.cfg`

**Comment by Xe-no1** [comment]
Yep I did that exact command, still doesn't recognise any snapshots

**Comment by Xe-no1** [comment]
> Try remaking grub, the automation might not be working. `sudo grub-mkconfig -o /boot/grub/grub.cfg`

Yep I did that exact command, still doesn't recognise any snapshots

**Comment by Schievel1** [comment]
Maybe you have forgotten to start the daemon? 

Try 
'systemctl status grub-btrfsd' to find out, and 'systemctl start grub-btrfsd' to start it if it isn't

**Comment by Xe-no1** [comment]
> Maybe you have forgotten to start the daemon?
> 
> Try 'systemctl status grub-btrfsd' to find out, and 'systemctl start grub-btrfsd' to start it if it isn't

I have used the 'systemctl enable grub-btrfsd' so it should start on system startup anyway, and after editing the config to support Timeshift I did 'systemctl restart grub-btrfsd'. I double checked anyway with 'systemctl status grub-btrfsd' and the grub-btrfsd is active

**Comment by Schievel1** [comment]
Yes that's sounds good to me. 

Can you watch the syslog (I think with systemd the command for that should be 'journalctl -f') then open timeshift, do a snapshot in timeshift and close it again? Then post the output of the syslog here.

**Comment by Xe-no1** [comment]
> Yes that's sounds good to me.
> 
> Can you watch the syslog (I think with systemd the command for that should be 'journalctl -f') then open timeshift, do a snapshot in timeshift and close it again? Then post the output of the syslog here.

Idk how to make the file show the contents automatically but this is the best I could do,
Here is the exact full output as a txt file: 
[syslog output.txt](https://github.com/user-attachments/files/16469528/syslog.output.txt)

**Comment by Schievel1** [comment]
So it seems like grub-btrfsd is working correctly. It detects snapshots being created and trigger the grub submenu creation (it says "grub menu recreated" in the logs after timeshift creates the snapshot)

What this does is running '/etc/grub.d/41_snapshots-btrfs' 
So let's continue there. Maybe you can run this command and see if that gives us any errors. Also, if you create a snapshot in timeshift and after that run this command, does the snapshot show up in the next boot?

**Comment by Xe-no1** [comment]
Okay to address your points, running the first command just says "Detecting snapshots..." on all occasions (Both before and after Timeshift snapshot creation), not an error but not a success either, Here is the output: [grub command output.txt](https://github.com/user-attachments/files/16475948/grub.command.output.txt), the first two were before making a snapshot and the third output is after the Timeshift snapshot, all of them are the exact same. No, no snapshot boot entries showed up in the next boot after making a snapshot and running the former command. As for the second command, the "btrfs subvolume list -sa /", here is the output of that command, I had to use 'sudo' for elevated privileges: [btrfs subvolume list output.txt](https://github.com/user-attachments/files/16475968/btrfs.subvolume.list.output.txt)

**Comment by Schievel1** [comment]
Sorry I forgot to mention, the first command also needs sudo

**Comment by Xe-no1** [comment]
> Sorry I forgot to mention, the first command also needs sudo

In that case, after running it with sudo I do get lots of text, and in the end it just ends up saying "No snapshots found". Here is the output: ```
[mazentech@Archie ~]$ sudo /etc/grub.d/41_snapshots-btrfs
[sudo] password for mazentech: 
Detecting snapshots ...
if [ ! -e "${prefix}/grub-btrfs.cfg" ]; then
echo ""
else
submenu 'Arch Linux ARM snapshots' {
    configfile "${prefix}/grub-btrfs.cfg"
}
fi
No snapshots found.
If you think an error has occurred, please file a bug report at "https://github.com/Antynea/grub-btrfs"
Unmount /tmp/grub-btrfs.whr3wpL0dT .. Success
[mazentech@Archie ~]$ sudo /etc/grub.d/41_snapshots-btrfs
Detecting snapshots ...
if [ ! -e "${prefix}/grub-btrfs.cfg" ]; then
echo ""
else
submenu 'Arch Linux ARM snapshots' {
    configfile "${prefix}/grub-btrfs.cfg"
}
fi
No snapshots found.
If you think an error has occurred, please file a bug report at "https://github.com/Antynea/grub-btrfs"
Unmount /tmp/grub-btrfs.uFMxYDLx1Y .. Success
[mazentech@Archie ~]$
```

**Comment by Schievel1** [comment]
Ok, so executing this script is what grub does as well. And it gets back this snippet you postet and integrates that into the grub.cfg file in /boot/grub. 
For some reason the script can't find any snapshots but btrfs-progs can. Hm. 

You say you're running grub-btrfs version 4.12. is there any particular reason for that? I know there have been some problems with changes in btrfs-progs in the past  which led to grub-btrfs not working anymore. Could you try upgrading to 4.13?

**Comment by Xe-no1** [comment]
> Ok, so executing this script is what grub does as well. And it gets back this snippet you postet and integrates that into the grub.cfg file in /boot/grub. For some reason the script can't find any snapshots but btrfs-progs can. Hm.
> 
> You say you're running grub-btrfs version 4.12. is there any particular reason for that? I know there have been some problems with changes in btrfs-progs in the past which led to grub-btrfs not working anymore. Could you try upgrading to 4.13?

Yes about that, I'm having the same issue as this guy here: #341. I just used 'pacman' to install grub-btrfs, not even the AUR
and it is supposed to be on 4.13, but as in the grub-btrfs config file it actually says that I have version 4.12

**Comment by Xe-no1** [comment]
To add on, if necessary I could compile from source and get the package that way, but idk if the grub-btrfs config is just wrong and that I _am_  using 4.13, or if the 'pacman' package is actually not up to date with the source on GitHub.

**Comment by Schievel1** [comment]
No you're fine. I think that was just a mistake I made in the 4.13 commit.

**Comment by Xe-no1** [comment]
Okay, so I am on the latest version. What you're saying so far is that btrfs-progs detects the snapshot but the script doesn't, yes?

**Comment by Schievel1** [comment]
Yes. The script is using btrfs-progs as well, then parses the output of btrfs-progs using awk  and such. 
If something in the output of btrfs-progs changes, the script will fail. But then again, if that is the case you wouldn't be the only one opening an issue here I guess.

**Comment by Xe-no1** [comment]
Another thing I must add, why is the output so crowded and 'unclean', I thought it would just say "Detecting snapshots" and directly after mention it finding any snapshots or not, but in the output I get, it looks very confusing and crowded.

**Comment by Schievel1** [comment]
You mean this part? 

if [ ! -e "${prefix}/grub-btrfs.cfg" ]; then
echo ""
else
submenu 'Arch Linux ARM snapshots' {
    configfile "${prefix}/grub-btrfs.cfg"
}
fi

This is the part grub integrates into its scripts. 
Those scripts work by generating output conditionally. Grub-mkconfig just calls the scripts in /etc/grub.d one by one and fetches whatever they output to stdout and puts it into /boot/grub/grub.cfg.
Usually you don't see this because grub-mkconfig redirects stdout of those scripts. But you can see the "detecting snapshots" and "no snapshots found" part because is printed to stderr. Grub-mkconfig doesn't redirect that. 

You said that you are new to Linux, you maybe want to read about stdin, stderr and stdout. It's a very powerful and old concept of Unix and the way it can chain programs together.

**Comment by Xe-no1** [comment]
> You mean this part?
> 
> if [ ! -e "${prefix}/grub-btrfs.cfg" ]; then echo "" else submenu 'Arch Linux ARM snapshots' { configfile "${prefix}/grub-btrfs.cfg" } fi
> 
> This is the part grub integrates into its scripts. Those scripts work by generating output conditionally. Grub-mkconfig just calls the scripts in /etc/grub.d one by one and fetches whatever they output to stdout and puts it into /boot/grub/grub.cfg. Usually you don't see this because grub-mkconfig redirects stdout of those scripts. But you can see the "detecting snapshots" and "no snapshots found" part because is printed to stderr. Grub-mkconfig doesn't redirect that. Since we ran the script manually without grub-mkconfig we can see it's whole output.
> 
> You said that you are new to Linux, you maybe want to read about stdin, stderr and stdout. It's a very powerful and old concept of Unix and the way it can chain programs together.

Okay good to know, will read about those, as I'm finding Linux and Unix very interesting so far and I would love to get even deeper. Thanks!

**Comment by Get-A-Pear-Today** [comment]
> There is one issue where the user had the same problem and with migration from timeshift to snapper it got solved. Instead of migrating directly, maybe make a manual snapshot first and see if it is detected. To do this run 'sudo mkdir /snapshots' 'btrfs subvolume snapshot / /snapshots/mysnapshot' Then run grub-config again as usual.
>

Hello! Similar to the OP I am also very new to Linux, even newer to Arch. I've been running into the same issue with having no snapshots detected and have been spending a lot of time looking through different posts for an answer with no solution working yet. I tried what you mentioned here and that did get 'sudo /etc/grub.d/41_snapshots-btrfs' to recognize 1 snapshot, but not my other 7 snapshots created by Timeshift. I'm confused as to where to go from here. Would changing where Timeshift creates snapshots to this new directory resolve this or is there more to it?

**Comment by ziomek3120** [comment]
I have exactly the same problem, after setting the --timeshift-auto parameter it does not detect snapshots. Btrfs progs detects them correctly. I use arch, I tested the version from the official repositories and from aur with the -git addendum, unfortunately the effect is the same, when executing grub-mkconfig it does not detect any snapshots. The command sudo /etc/grub.d/41_snapshots-btrfs also does not detect any snapshots. I have no idea what the problem could be.

**Comment by IncogCyberpunk** [comment]
Same problem here, the snapshots are created by Timeshift are present but they are just not being detected by the grub-btrfs. 
Any solution found ?

**Comment by Bthxtly** [comment]
Same issue, I find snapshots under `/timeshift-btrfs` directory, but `grub-btrfs` just can't find them.
```
❯ ls /timeshift-btrfs/snapshots
 2025-05-14_19-00-39   2025-05-18_23-00-00   2025-05-21_17-57-27
 2025-05-17_22-37-15   2025-05-19_18-49-09   2025-05-21_21-28-31
```
